lib/base/io-engine.cpp

```diff
@@ -16,60 +17,59 @@
 
 using namespace icinga;
 
-CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc)
+CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc, boost::asio::io_context::strand& strand)
 	: m_Done(false)
 {
-	auto& ioEngine (IoEngine::Get());
+	VERIFY(strand.running_in_this_thread());
+
+	auto& ie (IoEngine::Get());
+	Shared<AsioConditionVariable>::Ptr cv;
 
 	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+		for (auto freeSlots (ie.m_CpuBoundSemaphore.load(std::memory_order_relaxed)); freeSlots > 0;) {
+			if (ie.m_CpuBoundSemaphore.compare_exchange_weak(freeSlots, freeSlots - 1, std::memory_order_seq_cst, std::memory_order_relaxed)) {
+				return;
+			}
+		}
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
+		if (!cv) {
+			cv = Shared<AsioConditionVariable>::Make(ie.GetIoContext());
 			continue;
 		}
 
-		break;
-	}
-}
+		{
+			std::shared_lock lock (ie.m_OnOneFreeSlotMutex);
 
-CpuBoundWork::~CpuBoundWork()
-{
-	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
+			if (ie.m_CpuBoundSemaphore.load(std::memory_order_relaxed) > 0) {
+				continue;
+			}
+
+			ie.m_OnOneFreeSlot.connect([strand, cv] {
+				boost::asio::post(strand, [cv] { cv->NotifyOne(); });
+			});
+		}
+
+		cv->Wait(yc);
 	}
 }
 
 void CpuBoundWork::Done()
 {
 	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-
 		m_Done = true;
-	}
-}
 
-IoBoundWorkSlot::IoBoundWorkSlot(boost::asio::yield_context yc)
-	: yc(yc)
-{
-	IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-}
+		auto& ie (IoEngine::Get());
 
-IoBoundWorkSlot::~IoBoundWorkSlot()
-{
-	auto& ioEngine (IoEngine::Get());
+		if (ie.m_CpuBoundSemaphore.fetch_add(1) < 1) {
+			decltype(ie.m_OnOneFreeSlot) f;
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+			{
+				std::unique_lock lock (ie.m_OnOneFreeSlotMutex);
```

Comment:
AFAIK a lost wake-up takes place only if it happens before the to be waked up subscribes. But the (decision who to) wake-up is mutexed, so isn't it rather up to line 41?

Reply:
> But the (decision who to) wake-up is mutexed

With that, do you refer to the following block?

https://github.com/Icinga/icinga2/blob/3be2c3983f3a84e7815e587c006c74dff3565512/lib/base/io-engine.cpp#L66-L69

If so, that's inside the `if (ie.m_CpuBoundSemaphore.fetch_add(1) < 1)` which can decide that no one needs to be woken while another coroutine is currently in the process of subscribing itself to be woken and then goes to sleep.

Reply:
I don't see how `std::memory_order_seq_cst` is supposed to help with that (as claimed in the newly pushed changes):

https://github.com/Icinga/icinga2/blob/7db21b54c268e59880b06db23105bb214f7b261b/lib/base/io-engine.cpp#L52-L61

Reply:
Exactly! Done() can decide that no one needs to be woken, as: there are free slots available on the fast path, hence no one _would_ even subscribe to the slow path. The latter is true because both `ie.m_CpuBoundSemaphore.fetch_add(1) < 1` and `fastPath(std::memory_order_seq_cst)` sync with each other as strongly as possible. Maybe the explicit `std::memory_order_seq_cst` is even not needed here, let me know, but its purpose is to ensure either:

* `fastPath(std::memory_order_seq_cst)` succeeds. Lost wake-ups don't matter then. End of story.
* It fails and hence subscribes. This means m_CpuBoundSemaphore is 0 right now. Sure, latter could change in Done() after 1ns. Then it would wake up. But that is mutexed and has to wait for m_CpuBoundWaiting.emplace_back which still holds the mutex.

Reply:
I'm not yet sure what the role of `memory_order_seq_cst` is supposed to be here. Do you expect it to synchronize with the mutex in some way?

There is the critical section for subscribing to be woken: 
https://github.com/Icinga/icinga2/blob/7db21b54c268e59880b06db23105bb214f7b261b/lib/base/io-engine.cpp#L52-L61

Is there anything preventing that code from happening in between here (so at line 81)?
https://github.com/Icinga/icinga2/blob/7db21b54c268e59880b06db23105bb214f7b261b/lib/base/io-engine.cpp#L77-L84

Like this is not guaranteed by the use of the mutex alone, so this probably needs some higher-level argument that's not immediately obvious from the code.

Reply:
Please see the new/updated comments in both methods.